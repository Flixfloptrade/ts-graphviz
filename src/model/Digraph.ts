import { RootClusterType } from '../types';
import { RootCluster } from './RootCluster';
/**
 * Digraph object.
 *
 * @description
 * The object representing a directional graph.
 * @category Primary
 */
export class Digraph extends RootCluster {
  public readonly type = RootClusterType.digraph;
}
