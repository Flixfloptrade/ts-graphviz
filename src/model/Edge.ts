import { DotBase, IEdgeTarget, RootClusterType } from '../common';
import { commentOut, joinLines } from '../utils/dot-rendering';
import { Attributes } from './Attributes';
import { IContext } from './Context';
import { isEdgeTarget } from './Node';

/**
 * @category Primary
 */
export class Edge extends DotBase {
  /** Comments to include when outputting with toDot. */
  public comment?: string;
  /** @hidden */
  public readonly attributes = new Attributes();
  /** @hidden */
  private readonly targets: IEdgeTarget[];
  /** @hidden */
  private get dotArrow(): string {
    switch (this.context.graphType) {
      case RootClusterType.graph:
        return '--';
      case RootClusterType.digraph:
      default:
        return '->';
    }
  }

  constructor(context: IContext, target1: IEdgeTarget, target2: IEdgeTarget);
  constructor(context: IContext, ...targets: IEdgeTarget[]);
  constructor(private context: IContext, target1: IEdgeTarget, target2: IEdgeTarget, ...targets: IEdgeTarget[]) {
    super();
    this.targets = [target1, target2, ...targets].filter(n => isEdgeTarget(n));
  }

  /** Convert Edge to Dot language. */
  public toDot(): string {
    const comment = this.comment ? commentOut(this.comment) : undefined;
    const arrow = ` ${this.dotArrow} `;
    const target = this.targets.map(n => n.toEdgeTargetDot()).join(arrow);
    const attrs = this.attributes.size > 0 ? ` ${this.attributes.toDot()}` : '';
    const dot = `${target}${attrs};`;
    return joinLines(comment, dot);
  }
}
