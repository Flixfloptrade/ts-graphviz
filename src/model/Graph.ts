import { RootClusterType } from '../types';
import { RootCluster } from './RootCluster';
/**
 * Graph object.
 *
 * @description
 * An object representing an omnidirectional graph.
 * @category Primary
 */
export class Graph extends RootCluster {
  public readonly type = RootClusterType.graph;
}
