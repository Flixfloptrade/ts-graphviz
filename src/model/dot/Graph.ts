import { GraphAttributes } from '../values/attributes';
import { Dot, GraphType } from './Dot';
export class Graph extends Dot {
  public type: GraphType = 'graph';
  constructor(id: string = 'G', attributes: GraphAttributes = new GraphAttributes()) {
    super(id, attributes);
  }
}
