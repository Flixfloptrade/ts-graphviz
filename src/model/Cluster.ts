/* eslint-disable @typescript-eslint/no-use-before-define,max-classes-per-file */
import { attribute } from '../attribute';
import {
  ClusterType,
  EdgeTargetLike,
  ICluster,
  IClusterCommonAttributes,
  IEdge,
  INode,
  ISubgraph,
  Compass,
} from '../types';
import { AttributesBase } from './AttributesBase';
import { isEdgeTarget, ForwardRefNode, isEdgeTargetLike, Node } from './Node';
import { Edge } from './Edge';
import { IEdgeTarget } from '../types';
import { Attributes } from './Attributes';

/**
 * Base class for clusters.
 * @hidden
 */
export abstract class Cluster<T extends string> extends AttributesBase<T> implements ICluster<T> {
  /** Cluster ID */
  public id?: string;

  /** Comments to include when outputting with toDot. */
  public comment?: string;
  /** Indicates the type of cluster. */
  public abstract readonly type: ClusterType;
  /** Common attributes of objects in the cluster. */
  public abstract readonly attributes: Readonly<IClusterCommonAttributes>;

  /**
   * Nodes in the cluster.
   * @hidden
   */
  public nodes: Map<string, INode> = new Map();

  /**
   * Edges in the cluster.
   * @hidden
   */
  public edges: Set<IEdge> = new Set();

  /**
   * Subgraphs in the cluster.
   * @hidden
   */
  public subgraphs: Set<ISubgraph> = new Set();

  // private readonly internal = {
  //   nodes: ReadonlyMap
  // };

  /**
   * Add a Node to the cluster.
   */
  public addNode(node: INode): void {
    this.nodes.set(node.id, node);
  }

  /**
   * Add Edge to the cluster.
   */
  public addEdge(edge: IEdge): void {
    this.edges.add(edge);
  }

  /**
   * Add a Subgraph to the cluster.
   */
  public addSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.add(subgraph);
  }

  /**
   * Check if the Node exists in the cluster.
   */
  public existNode(nodeId: string): boolean {
    return this.nodes.has(nodeId);
  }

  /**
   * Check if the Edge exists in the cluster.
   */
  public existEdge(edge: IEdge): boolean {
    return this.edges.has(edge);
  }

  /**
   * Check if the Subgraph exists in the cluster.
   */
  public existSubgraph(subgraph: ISubgraph): boolean {
    return this.subgraphs.has(subgraph);
  }

  /**
   * Create a Subgraph and add it to the cluster.
   */
  public createSubgraph(id?: string): ISubgraph {
    const graph = new Subgraph(id);
    this.subgraphs.add(graph);
    return graph;
  }

  /**
   * Remove Node from the cluster.
   */
  public removeNode(node: INode | string): void {
    this.nodes.delete(typeof node === 'string' ? node : node.id);
  }

  /**
   * Remove Edge from the cluster.
   */
  public removeEdge(edge: IEdge): void {
    this.edges.delete(edge);
  }

  /**
   * Remove Subgraph from the cluster.
   */
  public removeSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.delete(subgraph);
  }

  /**
   * Create a Node in the cluster.
   */
  public createNode(id: string): INode {
    const node = new Node(id);
    this.nodes.set(id, node);
    return node;
  }

  /**
   * Get Subgraph in cluster by specifying id.
   *
   * If there is no Subgraph with the specified id in the cluster, return undefined.
   */
  public getSubgraph(id: string): ISubgraph | undefined {
    return Array.from(this.subgraphs.values()).find((subgraph) => subgraph.id === id);
  }

  /**
   * Get Node in cluster by specifying id.
   *
   * @description
   * If there is no Node with the specified id in the cluster, return undefined.
   */
  public getNode(id: string): INode | undefined {
    return this.nodes.get(id);
  }

  /** Create Edge and add it to the cluster. */
  public createEdge(targets: EdgeTargetLike[]): IEdge {
    if (targets.length < 2 && (isEdgeTargetLike(targets[0]) && isEdgeTargetLike(targets[1])) === false) {
      throw new Error('The element of Edge target is missing or not satisfied as Edge target.');
    }
    const edge = new Edge(targets.map((t) => this.toNodeLikeObject(t)));
    this.edges.add(edge);
    return edge;
  }

  /** @hidden */
  private toNodeLikeObject(node: EdgeTargetLike): IEdgeTarget {
    if (isEdgeTarget(node)) {
      return node;
    }
    const [id, port, compass] = node.split(':');
    const n = this.getNode(id);
    if (n !== undefined) {
      if (port && (compass === undefined || Compass.is(compass))) {
        return n.port({ port, compass });
      }
      return n;
    }
    if (Compass.is(compass)) {
      return new ForwardRefNode(id, { port, compass });
    }
    return new ForwardRefNode(id, { port });
  }

  /**
   * Declarative API for Subgraph.
   *
   * @description
   * If there is a Subgraph with the given ID, use it.
   * If not, create a Subgraph.
   *
   * @param id Subgraph ID.
   * @param callback Callback to operate Subgraph.
   */
  public subgraph(id?: string, callback?: (subgraph: ISubgraph) => void): ISubgraph {
    const subgraph: ISubgraph = id ? this.getSubgraph(id) ?? this.createSubgraph(id) : this.createSubgraph();
    if (callback) {
      callback(subgraph);
    }
    return subgraph;
  }

  /**
   * Declarative API for Node.
   *
   * @description
   * If there is a Node with the given ID, use it.
   * If not, create a Node.
   *
   * @param id Node ID.
   * @param callback Callback to operate Node.
   */
  public node(id: string, callback?: (node: INode) => void): INode {
    const node = this.getNode(id) ?? this.createNode(id);
    if (callback) {
      callback(node);
    }
    return node;
  }

  /**
   * Declarative API for Edge.
   *
   * @param targets Edges.
   * @param callback Callback to operate Edge.
   */
  public edge(targets: EdgeTargetLike[], callback?: (edge: IEdge) => void): IEdge {
    const edge = this.createEdge(targets);
    if (callback) {
      callback(edge);
    }
    return edge;
  }
}

/**
 * Subgraph object.
 * @category Primary
 */
export class Subgraph extends Cluster<attribute.Subgraph | attribute.ClusterSubgraph> implements ISubgraph {
  /** Indicates the type of cluster. */
  public type = ClusterType.subgraph;
  public attributes = {
    graph: new Attributes<attribute.ClusterSubgraph>(),
    edge: new Attributes<attribute.Edge>(),
    node: new Attributes<attribute.Node>(),
  };
  constructor(public readonly id?: string) {
    super();
  }
  /** Determines whether the Subgraph is a SubgraphCluster. */
  public isSubgraphCluster(): boolean {
    if (typeof this.id === 'string') {
      return this.id.startsWith('cluster');
    }
    return false;
  }
}
