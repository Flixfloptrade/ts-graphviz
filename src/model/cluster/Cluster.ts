import { DotBase } from '../../common';
import { SubgraphAttributes } from '../attributes';
import { Attributes } from '../attributes/Attributes';
import { EdgeAttributes } from '../attributes/EdgeAttributes';
import { NodeAttributes } from '../attributes/NodeAttributes';
import { Context } from '../Context';
import { DigraphEdge, Edge, GraphEdge } from '../Edge';
import { isNodeLike, isNodeLikeObject, Node, NodeLike, NodeLikeObject } from '../Node';

// tslint:disable: max-classes-per-file

export type RootClusterType = 'digraph' | 'graph';

export type ClusterType = RootClusterType | 'subgraph';
/**
 * @hidden
 */
export interface IClusterCommonAttributes<ATTR extends Attributes> {
  graph: ATTR;
  edge: EdgeAttributes;
  node: NodeAttributes;
}

/**
 * @hidden
 */
export abstract class Cluster<ATTR extends Attributes> extends DotBase {
  public abstract readonly context: Context;
  public abstract readonly type: ClusterType;
  public readonly attributes: Readonly<IClusterCommonAttributes<ATTR>>;

  private nodes: Map<string, Node> = new Map();

  private edges: Set<Edge> = new Set();

  private subgraphs: Map<string, Subgraph> = new Map();

  constructor(public readonly id: string, attributes: ATTR) {
    super();
    this.attributes = {
      graph: attributes,
      edge: new EdgeAttributes(),
      node: new NodeAttributes(),
    };
  }

  public add(object: Node | Edge | Subgraph): void {
    if (object instanceof Node) {
      this.addNode(object);
    } else if (object instanceof Edge) {
      this.addEdge(object);
    } else if (object instanceof Subgraph) {
      this.addSubgraph(object);
    }
  }

  public addNode(node: Node): void {
    this.nodes.set(node.id, node);
  }

  public addEdge(edge: Edge): void {
    this.edges.add(edge);
  }

  public addSubgraph(subgraph: Subgraph): void {
    this.subgraphs.set(subgraph.id, subgraph);
  }

  public existNode(nodeId: string): boolean {
    return this.nodes.has(nodeId);
  }

  public existEdge(edge: Edge): boolean {
    return this.edges.has(edge);
  }

  public existSubgraph(subgraphId: string): boolean {
    return this.subgraphs.has(subgraphId);
  }

  public createSubgraph(id: string): Subgraph {
    const graph = new Subgraph(this.context, id);
    this.subgraphs.set(id, graph);
    return graph;
  }

  public createNode(id: string): Node {
    const node = new Node(id);
    this.nodes.set(id, node);
    return node;
  }

  public getSubgraph(id: string): Subgraph | undefined {
    return this.subgraphs.get(id);
  }

  public getNode(id: string): Node | undefined {
    return this.nodes.get(id);
  }

  public createEdge(node1: NodeLike, node2: NodeLike, ...nodes: NodeLike[]): Edge;
  public createEdge(...nodes: NodeLike[]): Edge;
  public createEdge(node1: NodeLike, node2: NodeLike, ...nodes: NodeLike[]): Edge {
    if ((isNodeLike(node1) && isNodeLike(node2)) === false) {
      // TODO
      throw new Error();
    }

    const edge = new (this.context.graphType === 'graph' ? GraphEdge : DigraphEdge)(
      this.toNodeLikeObject(node1),
      this.toNodeLikeObject(node2),
      ...nodes.map(n => this.toNodeLikeObject(n)),
    );

    this.edges.add(edge);
    return edge;
  }

  public subgraph(id: string, callback: (subgraph: Subgraph) => void): Subgraph {
    const subgraph = this.getSubgraph(id) ?? this.createSubgraph(id);
    if (callback) {
      callback(subgraph);
    }
    return subgraph;
  }

  public node(id: string, callback?: (edge: Node) => void): Node {
    const node = this.getNode(id) ?? this.createNode(id);
    if (callback) {
      callback(node);
    }
    return node;
  }

  public edge(nodes: NodeLike[], callback?: (edge: Edge) => void): Edge {
    const edge = this.createEdge(...nodes);
    if (callback) {
      callback(edge);
    }
    return edge;
  }

  public toDot(): string {
    const type = this.type;
    const id = this.id;
    // attributes
    const commonAttributes = Object.entries(this.attributes)
      .filter(([_, attributes]) => attributes.size > 0)
      .map(([key, attributes]) => `${key} ${attributes.toDot()};`);

    // objects
    const nodes = Array.from(this.nodes.values()).map(o => o.toDot());
    const subgraphs = Array.from(this.subgraphs.values()).map(o => o.toDot());
    const edges = Array.from(this.edges.values()).map(o => o.toDot());
    const clusterContents = Cluster.joinLines(...commonAttributes, ...nodes, ...subgraphs, ...edges);
    const src = Cluster.joinLines(
      `${type} ${id} {`,
      clusterContents ? Cluster.indent(clusterContents) : undefined,
      '}',
    );
    return src;
  }

  private toNodeLikeObject(node: NodeLike): NodeLikeObject {
    if (isNodeLikeObject(node)) {
      return node;
    }
    // FIXME
    const [id, port] = node.split(':');
    const n = this.node(id);
    if (port) {
      return n.port(port);
    }
    return n;
  }
}

export abstract class RootCluster<ATTR extends Attributes> extends Cluster<ATTR> {
  public abstract readonly type: RootClusterType;
}

/**
 * @category Primary
 */
export class Subgraph extends Cluster<SubgraphAttributes> {
  public type: ClusterType = 'subgraph';

  constructor(public readonly context: Context, id: string, attributes: SubgraphAttributes = new SubgraphAttributes()) {
    super(id, attributes);
  }

  public isSubgraphCluster(): boolean {
    return this.id.startsWith('cluster_');
  }

  public toDot(): string {
    return `${super.toDot()};`;
  }
}
