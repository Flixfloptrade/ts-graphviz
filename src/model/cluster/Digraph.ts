import { RootClusterType } from '../../common';
import { RootCluster } from './Cluster';
/**
 * Digraph object.
 *
 * @description
 * The object representing a directional graph.
 * @category Primary
 */
export class Digraph extends RootCluster {
  public type = RootClusterType.digraph;
}
