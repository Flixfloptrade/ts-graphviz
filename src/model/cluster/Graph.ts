import { RootCluster, RootClusterType } from './Cluster';
/**
 * @category Primary
 */
export class Graph extends RootCluster {
  public type: RootClusterType = 'graph';
}
