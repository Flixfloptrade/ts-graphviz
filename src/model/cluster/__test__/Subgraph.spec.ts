import 'jest-graphviz';
import { DotBase, GraphvizObject } from '../../../common';
import { Edge } from '../../Edge';
import { Node } from '../../Node';
import { Cluster, Subgraph } from '../Cluster';
import { Digraph } from '../Digraph';
import { Graph } from '../Graph';

describe('class Subgraph', () => {
  let g: Digraph | Graph;
  let createEdge: (...nodes: Node[]) => Edge;
  const testCases: { title: string; beforeEachFunc: () => void }[] = [
    {
      title: 'root is Digraph',
      beforeEachFunc: () => {
        g = new Digraph();
        createEdge = (...nodes: Node[]) => new Edge({ graphType: 'graph' }, ...nodes);
      },
    },
    {
      title: 'root is Graph',
      beforeEachFunc: () => {
        g = new Graph();
        createEdge = (...nodes: Node[]) => new Edge({ graphType: 'graph' }, ...nodes);
      },
    },
  ];

  testCases.forEach(({ title, beforeEachFunc }) => {
    describe(title, () => {
      beforeEach(beforeEachFunc);
      let subgraph: Subgraph;

      beforeEach(() => {
        subgraph = g.context.createSubgraph('test');
      });

      it('should be instance of Subgraph/Cluster/DotBase/GraphvizObject', () => {
        expect(subgraph).toBeInstanceOf(Subgraph);
        expect(subgraph).toBeInstanceOf(Cluster);
        expect(subgraph).toBeInstanceOf(DotBase);
        expect(subgraph).toBeInstanceOf(GraphvizObject);
      });

      it('should be subgraph, when subgraph id is "test"', () => {
        subgraph = g.context.createSubgraph('test');
        expect(subgraph.isSubgraphCluster()).toBe(false);
      });

      it('should be escaped if id contains a newline character', () => {
        subgraph = new Subgraph(g.context);
        subgraph.id = '1\n2\n';
        const dot = g.toDot();
        expect(dot).toBeValidDotAndMatchSnapshot();
      });

      it('should be escaped if id contains a comma', () => {
        subgraph = new Subgraph(g.context);
        subgraph.id = '1"2"';
        const dot = g.toDot();
        expect(dot).toBeValidDotAndMatchSnapshot();
      });

      it('should be subgraph cluster, when subgraph id is "cluster_test"', () => {
        subgraph = g.context.createSubgraph('cluster_test');
        expect(subgraph.isSubgraphCluster()).toBe(true);
      });

      describe('label attribute behavior', () => {
        it('plain text label to be quoted by double quotation', () => {
          subgraph.attributes.graph.set('label', 'this is test for graph label');
          subgraph.attributes.edge.set('label', 'this is test for edge label');
          subgraph.attributes.node.set('label', 'this is test for node label');
          expect(subgraph.toDot()).toMatchSnapshot();
          expect(g.toDot()).toBeValidDot();
        });

        it('html like', () => {
          subgraph.attributes.graph.set('label', '<<B>this is test for graph label</B>>');
          subgraph.attributes.edge.set('label', '<<U>this is test for edge label</U>>');
          subgraph.attributes.node.set('label', '<<I>this is test for node label</I>>');
          expect(subgraph.toDot()).toMatchSnapshot();
          expect(g.toDot()).toBeValidDot();
        });
      });

      describe('addXxx existXxx removeXxx APIs', () => {
        it('Node operation methods works', () => {
          const id = 'node';
          expect(subgraph.existNode(id)).toBe(false);
          const node = new Node(id);
          subgraph.addNode(node);
          expect(subgraph.existNode(id)).toBe(true);
          subgraph.removeNode(node);
          expect(subgraph.existNode(id)).toBe(false);
          subgraph.addNode(node);
          expect(subgraph.existNode(id)).toBe(true);
          subgraph.removeNode(node.id);
          expect(subgraph.existNode(id)).toBe(false);
        });

        it('Edge operation methods works', () => {
          const [node1, node2] = ['node1', 'node2'].map(id => subgraph.createNode(id));
          const edge = createEdge(node1, node2);
          expect(subgraph.existEdge(edge)).toBe(false);
          subgraph.addEdge(edge);
          expect(subgraph.existEdge(edge)).toBe(true);
          expect(subgraph.toDot()).toMatchSnapshot();
          subgraph.removeEdge(edge);
          expect(subgraph.existEdge(edge)).toBe(false);
        });

        it('Subgraph operation methods works', () => {
          const sub = subgraph.context.createSubgraph('sub');
          expect(subgraph.existSubgraph(sub)).toBe(false);
          subgraph.addSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(true);
          expect(subgraph.toDot()).toMatchSnapshot();
          subgraph.removeSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(false);
          subgraph.addSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(true);
          subgraph.removeSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(false);
        });
      });
    });
  });
});
