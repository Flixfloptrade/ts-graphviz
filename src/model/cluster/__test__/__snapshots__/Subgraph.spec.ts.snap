// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Subgraph root is Digraph addXxx existXxx APIs Edge operation methods works 1`] = `
"subgraph \\"test\\" {
  \\"node1\\";
  \\"node2\\";
  \\"node1\\" -> \\"node2\\";
};"
`;

exports[`class Subgraph root is Digraph addXxx existXxx APIs Node operation methods works 1`] = `
"subgraph \\"test\\" {
  \\"node\\";
};"
`;

exports[`class Subgraph root is Digraph addXxx existXxx APIs Subgraph operation methods works 1`] = `
"subgraph \\"test\\" {
  subgraph \\"sub\\" {
  };
};"
`;

exports[`class Subgraph root is Digraph label attribute behavior html like 1`] = `
"subgraph \\"test\\" {
  graph [
    label = <<B>this is test for graph label</B>>,
  ];
  edge [
    label = <<U>this is test for edge label</U>>,
  ];
  node [
    label = <<I>this is test for node label</I>>,
  ];
};"
`;

exports[`class Subgraph root is Digraph label attribute behavior plain text label to be quoted by double quotation 1`] = `
"subgraph \\"test\\" {
  graph [
    label = \\"this is test for graph label\\",
  ];
  edge [
    label = \\"this is test for edge label\\",
  ];
  node [
    label = \\"this is test for node label\\",
  ];
};"
`;

exports[`class Subgraph root is Digraph should be escaped if id contains a comma 1`] = `
"digraph \\"G\\" {
}"
`;

exports[`class Subgraph root is Digraph should be escaped if id contains a newline character 1`] = `
"digraph \\"G\\" {
}"
`;

exports[`class Subgraph root is Graph addXxx existXxx APIs Edge operation methods works 1`] = `
"subgraph \\"test\\" {
  \\"node1\\";
  \\"node2\\";
  \\"node1\\" -- \\"node2\\";
};"
`;

exports[`class Subgraph root is Graph addXxx existXxx APIs Node operation methods works 1`] = `
"subgraph \\"test\\" {
  \\"node\\";
};"
`;

exports[`class Subgraph root is Graph addXxx existXxx APIs Subgraph operation methods works 1`] = `
"subgraph \\"test\\" {
  subgraph \\"sub\\" {
  };
};"
`;

exports[`class Subgraph root is Graph label attribute behavior html like 1`] = `
"subgraph \\"test\\" {
  graph [
    label = <<B>this is test for graph label</B>>,
  ];
  edge [
    label = <<U>this is test for edge label</U>>,
  ];
  node [
    label = <<I>this is test for node label</I>>,
  ];
};"
`;

exports[`class Subgraph root is Graph label attribute behavior plain text label to be quoted by double quotation 1`] = `
"subgraph \\"test\\" {
  graph [
    label = \\"this is test for graph label\\",
  ];
  edge [
    label = \\"this is test for edge label\\",
  ];
  node [
    label = \\"this is test for node label\\",
  ];
};"
`;

exports[`class Subgraph root is Graph should be escaped if id contains a comma 1`] = `
"graph \\"G\\" {
}"
`;

exports[`class Subgraph root is Graph should be escaped if id contains a newline character 1`] = `
"graph \\"G\\" {
}"
`;
